### PR TITLE
Store: Settings: Warn of Unsaved Changes if Navigating

### LINKED ISSUE #17271: Store: don't allow leaving Store settings unconfirmed and unsaved
When merchant changes his store settings it is still needed to confirm with `Save & Finish` button at the top of the page. It is possible though to leave the settings page ( clicking for example Orders in the left-hand side menu ) without clicking `Save & Finish`. This creates a possible flow that leaves merchant thinking that he has introduced changes in his settings but that is in fact not true. Leaving the settings page without clicking `Save & Finish` should show confirm or decline modal.

### PR DESCRIPTION
**Pull request by allendav**
Partially addresses #17271 (Payments and Taxes only)

WORK IN PROGRESS

To test:
```
* Navigate to Store > Settings > Payments
* Change your currency setting but don't save the change
* Click on a sidebar item (e.g. Reviews)
* Verify you are prompted that you have unsaved changes
* Hit Cancel

* Put the currency setting back the way it was
* Click on a sidebar item (e.g. Reviews)
* Verify you are NOT prompted that you have unsaved changes

* Change another payments setting, e.g. Your PayPal ID (email address for PayPal), Title for Check Payments, Test Secret Key for Stripe, etc etc.
* Click on a sidebar item (e.g. Reviews)
* Verify you are prompted that you have unsaved changes
* Hit Cancel
* (Note: Changing the store location does not require Save - it always saves immediately)

* Navigate to Store > Settings > Taxes
* If your store has the TaxJar plugin enabled, disable it
* Change one of the tax settings
* Click on a sidebar item (e.g. Reviews)
* Verify you are prompted that you have unsaved changes
* Hit Cancel

* Put the tax setting back the way it was
* Click on a sidebar item (e.g. Reviews)
* Verify you are NOT prompted that you have unsaved changes
```

Looks like this:

screen shot 2017-11-22 at 12 23 59 pm

Note: The payments approach involves testing the non-emptiness of the actions list to determine if the payment settings form has unsaved changes. The taxes approach involves testing the initial state against the state to which the controls are bound to determine if the tax settings form has unsaved changes.

### CODE REVIEW

File: client/extensions/woocommerce/app/settings/shipping/save-button.js [hunk 1]
```diff
@@ -82,38 +88,50 @@ class ShippingSettingsSaveButton extends Component {
...
-			) : null;
+		if ( finishedInitialSetup && ! wcsEnabled ) {
+			return null;
 		}
-		const label = wcsEnabled ? translate( 'Save and finish' ) : translate( "I'm Finished" );
+
+		const clickHandler = finishedInitialSetup ? this.save : this.redirect;
+		let label = translate( 'Save' );
+		if ( ! finishedInitialSetup ) {
+			label = wcsEnabled ? translate( 'Save and finish' ) : translate( "I'm Finished" );
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/settings/shipping/save-button.js]: Hi! I've found a possible matching string that has already been translated **37** times:
`translate( 'Finished!' )` *ES Score: 8*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/17271-prompt-when-unsaved-changes)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by allendav** [comment]
Switching gears for a moment to fix https://github.com/Automattic/wp-calypso/issues/20138 and then will proceed with this PR

**Comment by timmyc** [comment]
@allendav is this one ready for review now?

**Comment by allendav** [comment]
@timmyc - yes! ty!

File: client/extensions/woocommerce/state/data-layer/ui/payments/index.js [hunk 2]
```diff
@@ -4,124 +4,14 @@
...
 
-import { translate } from 'i18n-calypso';
-import { isEmpty, isEqual } from 'lodash';
+import { isEmpty } from 'lodash';
 
 /**
  * Internal dependencies
  */
 import { WOOCOMMERCE_PAYMENT_ACTION_LIST_CREATE } from 'woocommerce/state/action-types';
-import {
```
Comment (timmyc) [inline client/extensions/woocommerce/state/data-layer/ui/payments/index.js]: Curious why are all these changes in this PR?

Comment (allendav) [reply client/extensions/woocommerce/state/data-layer/ui/payments/index.js]: A little bit `prettier` but mostly to split the `getSaveSettingsActionListSteps` reducer out as a separate importable item

**Comment by allendav** [comment]
@timmyc wrote:

> When trying to actually save changes to a payment setting on this branch, I'm getting the following error

Thank you for catching that. Fixed in https://github.com/Automattic/wp-calypso/pull/20126/commits/54afa96d744cf89eac302a68f1e786a3ea2ae63e

**Comment by allendav** [comment]
@timmyc - omg looking :)

**Comment by allendav** [comment]
Abandoning this PR for a simpler approach that merely looks for whether an edit has occurred since the last save. https://github.com/Automattic/wp-calypso/pull/20930

**Pull request closed without merging**